catalog/core/catalog-core-backupplugin/src/main/java/ddf/catalog/backup/PeriodicBatchExecutor.java

```diff
@@ -0,0 +1,133 @@
+/**
+ * Copyright (c) Codice Foundation
+ * <p/>
+ * This is free software: you can redistribute it and/or modify it under the terms of the GNU Lesser
+ * General Public License as published by the Free Software Foundation, either version 3 of the
+ * License, or any later version.
+ * <p/>
+ * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without
+ * even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details. A copy of the GNU Lesser General Public License
+ * is distributed along with this program and can be found at
+ * <http://www.gnu.org/licenses/lgpl.html>.
+ **/
+
+package ddf.catalog.backup;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Consumer;
+
+public class PeriodicBatchExecutor<T> {
+
+    List<T> items;
+
+    private ExecutorService executor;
+
+    private ScheduledExecutorService scheduledExecutor;
+
+    private Consumer<List<T>> task;
+
+    private int chunkSize;
+
+    private long period;
+
+    private TimeUnit timeUnit;
+
+    private BlockingQueue<Future> futures;
+
+    public PeriodicBatchExecutor(int chunkSize, long period, TimeUnit timeUnit) {
+        if (chunkSize < 1) {
+            throw new IllegalArgumentException("Parameter chunkSize must be greater than zero");
+        }
+        if (period < 1) {
+            throw new IllegalArgumentException("Parameter period must be greater than zero ");
+        }
+        this.chunkSize = chunkSize;
+        this.period = period;
+        this.timeUnit = timeUnit;
+        this.executor = Executors.newSingleThreadExecutor();
+        this.scheduledExecutor = Executors.newSingleThreadScheduledExecutor();
+        this.futures = new LinkedBlockingQueue<>();
+        this.items = new ArrayList<>();
+    }
+
+    private synchronized void drain() {
+        for (List<T> batch : getBatches()) {
+            boolean success = futures.offer(executor.submit(() -> getTask().accept(batch)));
+            if(!success) {
+                throw new RuntimeException("Future object could not be added to the queue");
```

Comment:
It's usually bad practice to throw a raw `RuntimeException`. We should use`futures.add()` since we're not using a capacity-restricted queue, and throw something like `IllegalStateException` if `!success`.